lib/async_hooks.js

```diff
@@ -207,6 +208,84 @@ class AsyncResource {
   }
 }
 
+function getResource(id) {
+  const weak = idResourcePairing.get(id);
+  if (!weak) {
+    return undefined;
+  }
+  return weak.get();
+}
+
+function getCurrentResource() {
+  const id = executionAsyncId();
+  return getResource(id);
+}
+
+// AsyncContext part
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    contexts.forEach((context) => {
+      context._propagate(resource, getResource(triggerAsyncId));
+    });
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
```

Comment:
Some libraries may implement misbehaving `AsyncResource`s for which the `destroy` callback won't be ever called (just because these libraries don't call `emitDestroy`). [`FinalizationGroup` API](https://github.com/tc39/proposal-weakrefs#using-weakrefs-and-finalizationgroups-together) could be used to clean up garbage entries in `idResourcePairing` in such cases. But it's experimental and I'm not sure if it's available in node core's internal APIs or not.